### ISSUE TITLE
 Receive attachment emails in IMAP agent

### ISSUE THREAD

**Issue by bhagyalakshmikk** [cantino/huginn#1577]
I am working with IMAP agent , which  monitor incoming mails and forward it to a webservice . Is it possible to receive and forward an attachment mail to webservice ??

**Comment by dsander** [comment]
Hi @bhagyalakshmikk,

that is not possible at the moment. We have the concept of "file_pointers", but right now those Agents supporting it can only consume (process) the received files and turn the processed data into events. Allowing Agents to forward data from a source to a destination Agent would be really neat but would require some additional work.

**Comment by bhagyalakshmikk** [comment]
How can I achieve the concept of  "file_pointers"   ??

**Comment by jamsheer-thottathil** [comment]
How to forward a content of a attachment from imap to post agaent as event or payload

**Comment by dsander** [comment]
@bhagyalakshmikk The IMAP Agent would need the ability to emit file_pointers like for example the the [LocalFileAgent](https://github.com/cantino/huginn/blob/master/app/models/agents/local_file_agent.rb#L106), the PostAgent would then need to [consume](https://github.com/cantino/huginn/blob/master/app/models/agents/read_file_agent.rb#L46) the file_pointer and post the data to the website.

@jamsheer-thottathil Like I said in the previous response that is not possible at the moment.

**Comment by taenzerme** [comment]
@dsander Just so I understand this correctly:

I fetch mails with the IMAP agent and send these to another agent (POST). 

There is currently NO way to use the attachments in the mail? I need to get the file data from the mail body part, i.e. filename, content-type and the file content as string and want to POST this as a json payload to our web service.

I expected I could somehow access the other parts beside the body and parse them using regex to filter the parts.

What is the use of has_attachments then?

**Comment by cantino** [comment]
@taenzerme No, it doesn't look like the IMAPFolderAgent can currently emit the contents of attachments. That would be a reasonable addition if you'd like to add it.

**Comment by taenzerme** [comment]
@cantino I see, yes. I already did some experiments in the IMAPFolderAgent extracting the attachments, which is working. Currently I save the files to the systems tmp folder using the session id and pass a list of filenames to the payload.

Doing this right would require to extend the POSTAgent as well to handle files in an event payload.

Maybe I'll find the time this weekend.

Cheers

**Comment by neethkrish** [comment]
Hi @taenzerme ,

Could you please tell me how to make use of  IMAPFolderAgent in extracting the attachments and saving them to some folder. Also making use of POSTAgent to handle such saved files. 

Thanks in advance.

**Comment by taenzerme** [comment]
Hello @neethkrish, actually we wrote our own ImapAttachmentAgent to implement what was needed. At the moment there is no documentation but I attached the agent and our modifications to system_mailer.rb to this post. Maybe it helps getting started.

[huginn_imap_attachments.zip](https://github.com/huginn/huginn/files/1048624/huginn_imap_attachments.zip)
